Added a short-term fix for string literals used in tfop attrs

mhong opened the pull request:
String literals (used as tfop attrs) are again getting different SIL code pattern, causing the infamous error message:

```
/usr/local/google/home/hongm/ssd_part/git/swift-tensorflow-merge/swift/test/TensorFlowRuntime/tensor.swift:263:13: error: attribute 'padding' requires a constant argument
  let y = x.convolved2D(withFilter: filter, strides: (1, 1, 1, 1),
```

Filed https://bugs.swift.org/browse/SR-8117 for it, which should go away once we
move the partition pass to be right after deabstraction.

The short-term fix here is to avoid using small strings, so that compiler
will *not* generate the _StringObject.init(_smallUTF8SecondWord:) call.

This is done by disabling (comment out) the small string optimization code path in the 4 lines of code here https://github.com/apple/swift/blob/ad7cbb3a9324ff3ec99063f2174294feae420c28/stdlib/public/core/String.swift#L832-L835. 

Also tried an alternative suggested by @rxwei , where we pad the string literal
at the end of deabstraction pass, and remove the padding when lowering the insts
to TF ops. Unfortunately it did not work out -- for device string literals, even
after we pad with 16+ bytes (e.g. pad `/device:CPU:0` into
`________________/device:CPU:0`), `_StringObject.init(_smallUTF8SecondWord:)` is
still generated somehow.

Also renamed the device pseudo-attr from "device" to "__device", for consistency
with "__shapes".

The only test failure is now TensorFlow/tensor_autodiff.swift.

Commit:
String literals (used as tfop attrs) are again getting different SIL code
pattern, causing the infamous
/usr/local/google/home/hongm/ssd_part/git/swift-tensorflow-merge/swift/test/TensorFlowRuntime/tensor.swift:263:13: error: attribute 'padding' requires a constant argument
  let y = x.convolved2D(withFilter: filter, strides: (1, 1, 1, 1),

Filed https://bugs.swift.org/browse/SR-8117 for it, which should go away once we
move the partition pass to be right after deabstraction.

The short-term fix here is to avoid using small strings, so that compiler
will *not* generate the _StringObject.init(_smallUTF8SecondWord:) call.

Also tried an alternative suggested by rxwei@, where we pad the string literal
at the end of deabstraction pass, and remove the padding when lowering the insts
to TF ops. Unfortunately it did not work out -- for device string literals, even
after we pad with 16+ bytes (e.g. pad "/device:CPU:0" into
"________________/device:CPU:0"), _StringObject.init(_smallUTF8SecondWord:) is
still generated somehow.

Also renamed the device pseudo-attr from "device" to "__device", for consistency
with "__shapes".

The only compiler test failure is now TensorFlow/tensor_autodiff.swift.

mhong commented:
@swift-ci please test tensorflow

rxwei commented:
Will this block any usage of existing ops?

mhong commented:
The returned CI status is "Build finished. No test results found.". More details at https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/244/.

I'll try rerunning it.

mhong commented:
@rxwei Temporary blockage is possible. I suggest we push for getting deabstraction done and moving partitioning pass up ASAP.

mhong commented:
@swift-ci please test tensorflow

rxwei commented:
If this really ends up blocking conv2d op, then I suggest the merge should be blocked on full deabstraction. Otherwise conv nets won’t be possible.

Commit:
After doing a
./utils/update-checkout --scheme tensorflow --skip-repository swift

tensorflow stdlib no longer compiles. This commit fixes it.

In update-checkout-config.json, we currently have:
                "tensorflow-swift-bindings": "master"

We may need to version-control it with a commit hash. Same question for
                "tensorflow": "master",

mhong commented:
From local testing, this test does not build (it's expected in this branch).

/usr/local/google/home/hongm/ssd_part/git/swift-tensorflow-merge/swift/test/TensorFlowRuntime/tensor.swift:263:13: error: attribute 'padding' requires a constant argument
  let y = x.convolved2D(withFilter: filter, strides: (1, 1, 1, 1),
          ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/246/console also shows the build fails, but https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/246/ does not show any problem.

 @marcrasi @m3bm3b Is this an issue with our CI setup?

m3bm3b commented:
I am confused by several things here.

First, the error message you quote (tensor.swift:263:13: error: attribute 'padding' ...)
does not appear in the console output for the failing run.
Rather the error message in that log is
    Ops.swift:476:24: error: extraneous argument label 'values:'

Second, the console log says at the end:
   Recording test results
   ERROR: Step ‘Publish JUnit test result report’ failed: Test reports were found but none of them are new. Did tests run? 
   For example, /home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/Ninja-ReleaseAssert+stdlib- 
   Release/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/lit-tests.xml is 1 hr 43 min old

Third, to be more precise about 
   https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/246/
It's not that it does not show any problem; rather it
it indicate that there was some problem (it's red), but
it can't identify what the problem was.
This may be related to the second point.

But I don't know much about this build, or how it tries to identify problems.

mhong commented:
@m3bm3b, thanks for the investigation! The error in the first point is now fixed as part of this PR: https://github.com/apple/swift/pull/17546/commits/01cabf28014bfbb96da70958a35878eccdd069dc. I'll kick off a new CI run, and expect it to hit the issue "error: attribute 'padding' requires a constant argument" this time.

On https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/246/, it said "No identified problem" -- the color red helps, but is it possible to make the text more explicit that some error has occurred?

mhong commented:
@swift-ci please test tensorflow

mhong commented:
@rxwei, ack on this blocking conv2d. I'll give it another try in getting it to work.

Given we've all being investing a lot in this tensorflow-merge branch, it'd be great if we can merge it back into tensorflow branch ASAP to capitalize on the work. That'll also minimize the pains in future merging efforts (e.g. merge conflicts caused by large-scale code changes like getSwiftRValueType() -> getASTType() in this branch).

m3bm3b commented:
> the color red helps, but is it possible to make the text more explicit that some error has occurred?

Remember that this is Apple swift CI framework.
Mishal may be able to adjust things to improve this.   
I don't think that we have any direct control.  At least, I know of no config files to adjust.

mhong commented:
Update from the new run https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/248/:

The result page is clearly showing two test failures, and I'm pretty happy with it.
```
Test Result (2 failures / +2)Swift(linux-x86_64)
.Syntax.round_trip_parse_gen.swiftSwift(linux-x86_64)
.TensorFlow.tensor_autodiff.swift
```

So the above issue probably only occurs when there's a build failure, and that should be rare (people will usually build locally before triggering CI). Given Mike's input, we can sit on it for now and bring it up with Mishal later if this is still a pain point.

rxwei commented:
> @rxwei, ack on this blocking conv2d. I'll give it another try in getting it to work.

SG. Thanks for working on this! 

I'll take a look at the AD failure later this week.

mhong commented:
Chris, it'd be great to get your help. Thanks!

Test case (compiler only):

```swift
// RUN: %target-swift-frontend -Xllvm -tf-dump-intermediates -Xllvm -tf-dump-graph -O -emit-sil -verify %s | %FileCheck %s
import TensorFlow

public func testConvolution(x : Tensor<Float>, filter: Tensor<Float>) -> Tensor<Float> {
  return x.toAccelerator().convolved2D(withFilter: filter.toAccelerator(),
                       strides: (1, 2, 3, 4), padding: .same)
}
```

deabstraction result
```
--- TFDeabstraction Result: $S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF
// testConvolution(x:filter:)
sil @$S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF : $@convention(thin) (@guaranteed Tensor<Float>, @guaranteed Tensor<Float>) -> @owned Tensor<Float> {
// %0                                             // users: %4, %2
// %1                                             // users: %15, %3
bb0(%0 : $Tensor<Float>, %1 : $Tensor<Float>):
  debug_value %0 : $Tensor<Float>, let, name "x", argno 1 // id: %2
  debug_value %1 : $Tensor<Float>, let, name "filter", argno 2 // id: %3
  %4 = struct_extract %0 : $Tensor<Float>, #Tensor.handle // user: %6
  // function_ref __tf_send
  %5 = function_ref @__tf_send : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // user: %6
  %6 = apply %5<Float>(%4) : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // users: %131, %109, %99, %10, %9
  %7 = alloc_stack $Tensor<Float>                 // users: %8, %11, %14
  %8 = struct_element_addr %7 : $*Tensor<Float>, #Tensor.handle // user: %9
  store %6 to %8 : $*TensorHandle<Float>          // id: %9
  strong_retain %6 : $TensorHandle<Float>         // id: %10
  %11 = load %7 : $*Tensor<Float>                 // user: %12
  %12 = struct_extract %11 : $Tensor<Float>, #Tensor.handle // user: %13
  strong_release %12 : $TensorHandle<Float>       // id: %13
  dealloc_stack %7 : $*Tensor<Float>              // id: %14
  %15 = struct_extract %1 : $Tensor<Float>, #Tensor.handle // user: %17
  // function_ref __tf_send
  %16 = function_ref @__tf_send : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // user: %17
  %17 = apply %16<Float>(%15) : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // users: %130, %109, %100, %21, %20
  %18 = alloc_stack $Tensor<Float>                // users: %19, %22, %25
  %19 = struct_element_addr %18 : $*Tensor<Float>, #Tensor.handle // user: %20
  store %17 to %19 : $*TensorHandle<Float>        // id: %20
  strong_retain %17 : $TensorHandle<Float>        // id: %21
  %22 = load %18 : $*Tensor<Float>                // user: %23
  %23 = struct_extract %22 : $Tensor<Float>, #Tensor.handle // user: %24
  strong_release %23 : $TensorHandle<Float>       // id: %24
  dealloc_stack %18 : $*Tensor<Float>             // id: %25
  %26 = integer_literal $Builtin.Int32, 1         // user: %27
  %27 = struct $Int32 (%26 : $Builtin.Int32)      // user: %36
  %28 = integer_literal $Builtin.Int32, 2         // user: %29
  %29 = struct $Int32 (%28 : $Builtin.Int32)      // user: %36
  %30 = integer_literal $Builtin.Int32, 3         // user: %31
  %31 = struct $Int32 (%30 : $Builtin.Int32)      // user: %36
  %32 = integer_literal $Builtin.Int32, 4         // user: %33
  %33 = struct $Int32 (%32 : $Builtin.Int32)      // user: %36
  %34 = metatype $@thin Padding.Type
  %35 = enum $Padding, #Padding.same!enumelt      // user: %92
  %36 = tuple (%27 : $Int32, %29 : $Int32, %31 : $Int32, %33 : $Int32) // users: %89, %88, %87, %86, %82, %81, %80, %79, %75, %74, %73, %72, %68, %67, %66, %65
  %37 = integer_literal $Builtin.Word, 4          // users: %43, %38
  %38 = builtin "zextOrBitCast_Word_Int64"(%37 : $Builtin.Word) : $Builtin.Int64 // users: %41, %39
  %39 = struct $Int (%38 : $Builtin.Int64)        // users: %124, %46
  %40 = integer_literal $Builtin.Int64, 0         // user: %41
  %41 = builtin "cmp_slt_Int64"(%40 : $Builtin.Int64, %38 : $Builtin.Int64) : $Builtin.Int1 // user: %42
  cond_br %41, bb1, bb4                           // id: %42

bb1:                                              // Preds: bb0
  %43 = alloc_ref [tail_elems $Int32 * %37 : $Builtin.Word] $_ContiguousArrayStorage<Int32> // user: %46
  %44 = metatype $@thin Array<Int32>.Type         // user: %46
  // function_ref static Array._adoptStorage(_:count:)
  %45 = function_ref @$SSa13_adoptStorage_5countSayxG_SpyxGts016_ContiguousArrayB0CyxGn_SitFZ : $@convention(method) <τ_0_0> (@owned _ContiguousArrayStorage<τ_0_0>, Int, @thin Array<τ_0_0>.Type) -> (@owned Array<τ_0_0>, UnsafeMutablePointer<τ_0_0>) // user: %46
  %46 = apply %45<Int32>(%43, %39, %44) : $@convention(method) <τ_0_0> (@owned _ContiguousArrayStorage<τ_0_0>, Int, @thin Array<τ_0_0>.Type) -> (@owned Array<τ_0_0>, UnsafeMutablePointer<τ_0_0>) // users: %48, %47
  %47 = tuple_extract %46 : $(Array<Int32>, UnsafeMutablePointer<Int32>), 0 // user: %50
  %48 = tuple_extract %46 : $(Array<Int32>, UnsafeMutablePointer<Int32>), 1 // user: %49
  %49 = struct_extract %48 : $UnsafeMutablePointer<Int32>, #UnsafeMutablePointer._rawValue // user: %50
  br bb2(%47 : $Array<Int32>, %49 : $Builtin.RawPointer) // id: %50

// %51                                            // users: %56, %53
// %52                                            // user: %53
bb2(%51 : $Array<Int32>, %52 : $Builtin.RawPointer): // Preds: bb4 bb1
  %53 = tuple (%51 : $Array<Int32>, %52 : $Builtin.RawPointer) // users: %55, %54
  %54 = tuple_extract %53 : $(Array<Int32>, Builtin.RawPointer), 0
  %55 = tuple_extract %53 : $(Array<Int32>, Builtin.RawPointer), 1 // user: %56
  br bb3(%51 : $Array<Int32>, %55 : $Builtin.RawPointer) // id: %56

// %57                                            // users: %109, %59
// %58                                            // user: %59
bb3(%57 : $Array<Int32>, %58 : $Builtin.RawPointer): // Preds: bb2
  %59 = tuple (%57 : $Array<Int32>, %58 : $Builtin.RawPointer) // users: %60, %62, %63
  %60 = tuple_extract %59 : $(Array<Int32>, Builtin.RawPointer), 0 // users: %120, %102, %61
  retain_value %60 : $Array<Int32>                // id: %61
  %62 = tuple_extract %59 : $(Array<Int32>, Builtin.RawPointer), 1 // user: %64
  release_value %59 : $(Array<Int32>, Builtin.RawPointer) // id: %63
  %64 = pointer_to_address %62 : $Builtin.RawPointer to [strict] $*Int32 // users: %85, %78, %71, %69
  %65 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 0 // user: %69
  %66 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 1
  %67 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 2
  %68 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 3
  store %65 to %64 : $*Int32                      // id: %69
  %70 = integer_literal $Builtin.Word, 1          // user: %71
  %71 = index_addr %64 : $*Int32, %70 : $Builtin.Word // user: %76
  %72 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 0
  %73 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 1 // user: %76
  %74 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 2
  %75 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 3
  store %73 to %71 : $*Int32                      // id: %76
  %77 = integer_literal $Builtin.Word, 2          // user: %78
  %78 = index_addr %64 : $*Int32, %77 : $Builtin.Word // user: %83
  %79 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 0
  %80 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 1
  %81 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 2 // user: %83
  %82 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 3
  store %81 to %78 : $*Int32                      // id: %83
  %84 = integer_literal $Builtin.Word, 3          // user: %85
  %85 = index_addr %64 : $*Int32, %84 : $Builtin.Word // user: %90
  %86 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 0
  %87 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 1
  %88 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 2
  %89 = tuple_extract %36 : $(Int32, Int32, Int32, Int32), 3 // user: %90
  store %89 to %85 : $*Int32                      // id: %90
  // function_ref Padding.raw.getter
  %91 = function_ref @$S10TensorFlow7PaddingO3rawAA3RawOABOvg : $@convention(method) (Padding) -> Raw.Padding // user: %92
  %92 = apply %91(%35) : $@convention(method) (Padding) -> Raw.Padding // user: %104
  // function_ref default argument 3 of static Raw.conv2D<A>(_:filter:strides:useCudnnOnGpu:padding:dataFormat:dilations:)
  %93 = function_ref @$S10TensorFlow3RawO6conv2D_6filter7strides13useCudnnOnGpu7padding10dataFormat9dilationsAA0A0VyxGAM_AMSays5Int32VGSbAC7PaddingOAC04DataM0OAPts19BinaryFloatingPointRzAA013AccelerableByaB0RzlFZfA2_ : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> Bool // user: %94
  %94 = apply %93<Float>() : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> Bool // user: %108
  // function_ref default argument 5 of static Raw.conv2D<A>(_:filter:strides:useCudnnOnGpu:padding:dataFormat:dilations:)
  %95 = function_ref @$S10TensorFlow3RawO6conv2D_6filter7strides13useCudnnOnGpu7padding10dataFormat9dilationsAA0A0VyxGAM_AMSays5Int32VGSbAC7PaddingOAC04DataM0OAPts19BinaryFloatingPointRzAA013AccelerableByaB0RzlFZfA4_ : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> Raw.DataFormat // user: %96
  %96 = apply %95<Float>() : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> Raw.DataFormat // user: %106
  // function_ref default argument 6 of static Raw.conv2D<A>(_:filter:strides:useCudnnOnGpu:padding:dataFormat:dilations:)
  %97 = function_ref @$S10TensorFlow3RawO6conv2D_6filter7strides13useCudnnOnGpu7padding10dataFormat9dilationsAA0A0VyxGAM_AMSays5Int32VGSbAC7PaddingOAC04DataM0OAPts19BinaryFloatingPointRzAA013AccelerableByaB0RzlFZfA5_ : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> @owned Array<Int32> // user: %98
  %98 = apply %97<Float>() : $@convention(thin) <τ_0_0 where τ_0_0 : BinaryFloatingPoint, τ_0_0 : AccelerableByTensorFlow> () -> @owned Array<Int32> // users: %109, %119, %107
  strong_retain %6 : $TensorHandle<Float>         // id: %99
  strong_retain %17 : $TensorHandle<Float>        // id: %100
  %101 = metatype $@thick Float.Type              // user: %109
  retain_value %60 : $Array<Int32>                // id: %102
  // function_ref Raw.Padding.cName.getter
  %103 = function_ref @$S10TensorFlow3RawO7PaddingO5cNameSSvg : $@convention(method) (Raw.Padding) -> @owned String // user: %104
  %104 = apply %103(%92) : $@convention(method) (Raw.Padding) -> @owned String // user: %109
  // function_ref Raw.DataFormat.cName.getter
  %105 = function_ref @$S10TensorFlow3RawO10DataFormatO5cNameSSvg : $@convention(method) (Raw.DataFormat) -> @owned String // user: %106
  %106 = apply %105(%96) : $@convention(method) (Raw.DataFormat) -> @owned String // user: %109
  retain_value %98 : $Array<Int32>                // id: %107
  %108 = struct_extract %94 : $Bool, #Bool._value // user: %109
  %109 = builtin "__tfop_Conv2D,$in,$in,T,strides,use_cudnn_on_gpu,padding,data_format,dilations"(%6 : $TensorHandle<Float>, %17 : $TensorHandle<Float>, %101 : $@thick Float.Type, %57 : $Array<Int32>, %108 : $Builtin.Int1, %104 : $String, %106 : $String, %98 : $Array<Int32>) : $TensorHandle<Float> // users: %114, %112, %113
  %110 = alloc_stack $Tensor<Float>               // users: %111, %115, %118
  %111 = struct_element_addr %110 : $*Tensor<Float>, #Tensor.handle // user: %112
  store %109 to %111 : $*TensorHandle<Float>      // id: %112
  %113 = struct $Tensor<Float> (%109 : $TensorHandle<Float>) // user: %121
  strong_retain %109 : $TensorHandle<Float>       // id: %114
  %115 = load %110 : $*Tensor<Float>              // user: %116
  %116 = struct_extract %115 : $Tensor<Float>, #Tensor.handle // user: %117
  strong_release %116 : $TensorHandle<Float>      // id: %117
  dealloc_stack %110 : $*Tensor<Float>            // id: %118
  release_value %98 : $Array<Int32>               // id: %119
  release_value %60 : $Array<Int32>               // id: %120
  br bb5(%113 : $Tensor<Float>)                   // id: %121

bb4:                                              // Preds: bb0
  %122 = metatype $@thin Array<Int32>.Type        // user: %124
  // function_ref static Array._allocateUninitialized(_:)
  %123 = function_ref @$SSa22_allocateUninitializedySayxG_SpyxGtSiFZ : $@convention(method) <τ_0_0> (Int, @thin Array<τ_0_0>.Type) -> (@owned Array<τ_0_0>, UnsafeMutablePointer<τ_0_0>) // user: %124
  %124 = apply %123<Int32>(%39, %122) : $@convention(method) <τ_0_0> (Int, @thin Array<τ_0_0>.Type) -> (@owned Array<τ_0_0>, UnsafeMutablePointer<τ_0_0>) // users: %126, %125
  %125 = tuple_extract %124 : $(Array<Int32>, UnsafeMutablePointer<Int32>), 0 // user: %128
  %126 = tuple_extract %124 : $(Array<Int32>, UnsafeMutablePointer<Int32>), 1 // user: %127
  %127 = struct_extract %126 : $UnsafeMutablePointer<Int32>, #UnsafeMutablePointer._rawValue // user: %128
  br bb2(%125 : $Array<Int32>, %127 : $Builtin.RawPointer) // id: %128

// %129                                           // user: %132
bb5(%129 : $Tensor<Float>):                       // Preds: bb3
  strong_release %17 : $TensorHandle<Float>       // id: %130
  strong_release %6 : $TensorHandle<Float>        // id: %131
  return %129 : $Tensor<Float>                    // id: %132
} // end sil function '$S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF'
```

INPUT to tf partition pass:
```
---- INPUT FUNCTION $S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF ----------
// testConvolution(x:filter:)
sil @$S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF : $@convention(thin) (@guaranteed Tensor<Float>, @guaranteed Tensor<Float>) -> @owned Tensor<Float> {
// %0                                             // users: %4, %2
// %1                                             // users: %9, %3
bb0(%0 : $Tensor<Float>, %1 : $Tensor<Float>):
  debug_value %0 : $Tensor<Float>, let, name "x", argno 1 // id: %2
  debug_value %1 : $Tensor<Float>, let, name "filter", argno 2 // id: %3
  %4 = struct_extract %0 : $Tensor<Float>, #Tensor.handle // user: %6
  // function_ref __tf_send
  %5 = function_ref @__tf_send : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // users: %10, %6
  %6 = apply %5<Float>(%4) : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // users: %8, %108, %101, %60, %7
  strong_retain %6 : $TensorHandle<Float>         // id: %7
  strong_release %6 : $TensorHandle<Float>        // id: %8
  %9 = struct_extract %1 : $Tensor<Float>, #Tensor.handle // user: %10
  %10 = apply %5<Float>(%9) : $@convention(thin) <τ_0_0 where τ_0_0 : AccelerableByTensorFlow> (@guaranteed TensorHandle<τ_0_0>) -> @owned TensorHandle<τ_0_0> // users: %12, %107, %101, %61, %11
  strong_retain %10 : $TensorHandle<Float>        // id: %11
  strong_release %10 : $TensorHandle<Float>       // id: %12
  %13 = integer_literal $Builtin.Int32, 1         // user: %14
  %14 = struct $Int32 (%13 : $Builtin.Int32)      // users: %53, %55, %57, %59, %35
  %15 = integer_literal $Builtin.Int32, 2         // user: %16
  %16 = struct $Int32 (%15 : $Builtin.Int32)      // user: %38
  %17 = integer_literal $Builtin.Int32, 3         // user: %18
  %18 = struct $Int32 (%17 : $Builtin.Int32)      // user: %41
  %19 = integer_literal $Builtin.Int32, 4         // user: %20
  %20 = struct $Int32 (%19 : $Builtin.Int32)      // user: %44
  %21 = integer_literal $Builtin.Word, 4          // users: %46, %24
  %22 = integer_literal $Builtin.Int64, 4         // user: %23
  %23 = struct $Int (%22 : $Builtin.Int64)        // user: %28
  %24 = alloc_ref [tail_elems $Int32 * %21 : $Builtin.Word] $_ContiguousArrayStorage<Int32> // user: %25
  %25 = upcast %24 : $_ContiguousArrayStorage<Int32> to $_ContiguousArrayStorageBase // users: %33, %34, %62, %106, %32, %70, %30
  %26 = integer_literal $Builtin.Int64, 8         // user: %27
  %27 = struct $UInt (%26 : $Builtin.Int64)       // user: %28
  %28 = struct $_SwiftArrayBodyStorage (%23 : $Int, %27 : $UInt) // user: %29
  %29 = struct $_ArrayBody (%28 : $_SwiftArrayBodyStorage) // users: %49, %31
  %30 = ref_element_addr %25 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %31
  store %29 to %30 : $*_ArrayBody                 // id: %31
  %32 = ref_tail_addr %25 : $_ContiguousArrayStorageBase, $Int32 // users: %35, %37, %40, %43
  strong_retain %25 : $_ContiguousArrayStorageBase // id: %33
  strong_release %25 : $_ContiguousArrayStorageBase // id: %34
  store %14 to %32 : $*Int32                      // id: %35
  %36 = integer_literal $Builtin.Word, 1          // users: %54, %37
  %37 = index_addr %32 : $*Int32, %36 : $Builtin.Word // user: %38
  store %16 to %37 : $*Int32                      // id: %38
  %39 = integer_literal $Builtin.Word, 2          // users: %56, %40
  %40 = index_addr %32 : $*Int32, %39 : $Builtin.Word // user: %41
  store %18 to %40 : $*Int32                      // id: %41
  %42 = integer_literal $Builtin.Word, 3          // users: %58, %43
  %43 = index_addr %32 : $*Int32, %42 : $Builtin.Word // user: %44
  store %20 to %43 : $*Int32                      // id: %44
  %45 = integer_literal $Builtin.Int1, -1         // users: %68, %101
  %46 = alloc_ref [tail_elems $Int32 * %21 : $Builtin.Word] $_ContiguousArrayStorage<Int32> // user: %47
  %47 = upcast %46 : $_ContiguousArrayStorage<Int32> to $_ContiguousArrayStorageBase // users: %51, %52, %100, %105, %50, %72, %48
  %48 = ref_element_addr %47 : $_ContiguousArrayStorageBase, #_ContiguousArrayStorageBase.countAndCapacity // user: %49
  store %29 to %48 : $*_ArrayBody                 // id: %49
  %50 = ref_tail_addr %47 : $_ContiguousArrayStorageBase, $Int32 // users: %53, %54, %56, %58
  strong_retain %47 : $_ContiguousArrayStorageBase // id: %51
  strong_release %47 : $_ContiguousArrayStorageBase // id: %52
  store %14 to %50 : $*Int32                      // id: %53
  %54 = index_addr %50 : $*Int32, %36 : $Builtin.Word // user: %55
  store %14 to %54 : $*Int32                      // id: %55
  %56 = index_addr %50 : $*Int32, %39 : $Builtin.Word // user: %57
  store %14 to %56 : $*Int32                      // id: %57
  %58 = index_addr %50 : $*Int32, %42 : $Builtin.Word // user: %59
  store %14 to %58 : $*Int32                      // id: %59
  strong_retain %6 : $TensorHandle<Float>         // id: %60
  strong_retain %10 : $TensorHandle<Float>        // id: %61
  strong_retain %25 : $_ContiguousArrayStorageBase // id: %62
  %63 = integer_literal $Builtin.Int64, 7         // user: %67
  %64 = metatype $@thick UInt.Type                // user: %65
  %65 = builtin "strideof"<UInt>(%64 : $@thick UInt.Type) : $Builtin.Word // user: %66
  %66 = builtin "zextOrBitCast_Word_Int64"(%65 : $Builtin.Word) : $Builtin.Int64 // user: %67
  %67 = builtin "cmp_slt_Int64"(%63 : $Builtin.Int64, %66 : $Builtin.Int64) : $Builtin.Int1 // user: %68
  %68 = builtin "int_expect_Int1"(%67 : $Builtin.Int1, %45 : $Builtin.Int1) : $Builtin.Int1 // user: %69
  cond_br %68, bb1, bb2                           // id: %69

bb1:                                              // Preds: bb0
  %70 = struct $_ContiguousArrayBuffer<Int32> (%25 : $_ContiguousArrayStorageBase) // user: %71
  %71 = struct $Array<Int32> (%70 : $_ContiguousArrayBuffer<Int32>) // user: %101
  %72 = struct $_ContiguousArrayBuffer<Int32> (%47 : $_ContiguousArrayStorageBase) // user: %73
  %73 = struct $Array<Int32> (%72 : $_ContiguousArrayBuffer<Int32>) // user: %101
  %74 = metatype $@thick Float.Type               // user: %101
  %75 = integer_literal $Builtin.Int64, 288230376151711744 // users: %79, %91, %76
  %76 = struct $UInt (%75 : $Builtin.Int64)       // users: %92, %81
  %77 = integer_literal $Builtin.Int64, 1162690899 // users: %85, %79, %87
  // function_ref specialized _SmallUTF8String._invariantCheck()
  %78 = function_ref @$Ss16_SmallUTF8StringV15_invariantCheckyyFTf4x_n : $@convention(thin) (Builtin.Int64, Builtin.Int64) -> () // users: %91, %79
  %79 = apply %78(%77, %75) : $@convention(thin) (Builtin.Int64, Builtin.Int64) -> ()
  // function_ref specialized _StringObject.init(_smallUTF8SecondWord:)
  %80 = function_ref @$Ss13_StringObjectV20_smallUTF8SecondWordABSu_tcfCTf4nd_n : $@convention(thin) (UInt) -> @owned _StringObject // users: %92, %81
  %81 = apply %80(%76) : $@convention(thin) (UInt) -> @owned _StringObject // users: %82, %88
  %82 = struct_extract %81 : $_StringObject, #_StringObject._object // users: %85, %86, %83
  strong_retain %82 : $Builtin.BridgeObject       // id: %83
  // function_ref specialized _StringGuts._invariantCheck()
  %84 = function_ref @$Ss11_StringGutsV15_invariantCheckyyFTf4x_n : $@convention(thin) (@guaranteed Builtin.BridgeObject, Builtin.Int64) -> () // users: %95, %85
  %85 = apply %84(%82, %77) : $@convention(thin) (@guaranteed Builtin.BridgeObject, Builtin.Int64) -> ()
  strong_release %82 : $Builtin.BridgeObject      // id: %86
  %87 = struct $UInt (%77 : $Builtin.Int64)       // user: %88
  %88 = struct $_StringGuts (%81 : $_StringObject, %87 : $UInt) // user: %89
  %89 = struct $String (%88 : $_StringGuts)       // user: %101
  %90 = integer_literal $Builtin.Int64, 1129793614 // users: %97, %95, %91
  %91 = apply %78(%90, %75) : $@convention(thin) (Builtin.Int64, Builtin.Int64) -> ()
  %92 = apply %80(%76) : $@convention(thin) (UInt) -> @owned _StringObject // users: %93, %98
  %93 = struct_extract %92 : $_StringObject, #_StringObject._object // users: %95, %96, %94
  strong_retain %93 : $Builtin.BridgeObject       // id: %94
  %95 = apply %84(%93, %90) : $@convention(thin) (@guaranteed Builtin.BridgeObject, Builtin.Int64) -> ()
  strong_release %93 : $Builtin.BridgeObject      // id: %96
  %97 = struct $UInt (%90 : $Builtin.Int64)       // user: %98
  %98 = struct $_StringGuts (%92 : $_StringObject, %97 : $UInt) // user: %99
  %99 = struct $String (%98 : $_StringGuts)       // user: %101
  strong_retain %47 : $_ContiguousArrayStorageBase // id: %100
  %101 = builtin "__tfop_Conv2D,$in,$in,T,strides,use_cudnn_on_gpu,padding,data_format,dilations"(%6 : $TensorHandle<Float>, %10 : $TensorHandle<Float>, %74 : $@thick Float.Type, %71 : $Array<Int32>, %45 : $Builtin.Int1, %89 : $String, %99 : $String, %73 : $Array<Int32>) : $TensorHandle<Float> // users: %104, %103, %102
  %102 = struct $Tensor<Float> (%101 : $TensorHandle<Float>) // user: %109
  strong_retain %101 : $TensorHandle<Float>       // id: %103
  strong_release %101 : $TensorHandle<Float>      // id: %104
  strong_release %47 : $_ContiguousArrayStorageBase // id: %105
  strong_release %25 : $_ContiguousArrayStorageBase // id: %106
  strong_release %10 : $TensorHandle<Float>       // id: %107
  strong_release %6 : $TensorHandle<Float>        // id: %108
  return %102 : $Tensor<Float>                    // id: %109

bb2:                                              // Preds: bb0
  %110 = integer_literal $Builtin.Int8, 2         // users: %125, %119, %115
  %111 = integer_literal $Builtin.Int32, 0        // user: %126
  %112 = string_literal utf8 "/usr/local/google/home/hongm/ssd_part/git/swift-tensorflow-merge/swift/stdlib/public/core/SmallString.swift" // user: %114
  %113 = integer_literal $Builtin.Word, 107       // user: %115
  %114 = builtin "ptrtoint_Word"(%112 : $Builtin.RawPointer) : $Builtin.Word // user: %115
  %115 = struct $StaticString (%114 : $Builtin.Word, %113 : $Builtin.Word, %110 : $Builtin.Int8) // user: %128
  %116 = string_literal utf8 ""                   // user: %118
  %117 = integer_literal $Builtin.Word, 0         // user: %119
  %118 = builtin "ptrtoint_Word"(%116 : $Builtin.RawPointer) : $Builtin.Word // user: %119
  %119 = struct $StaticString (%118 : $Builtin.Word, %117 : $Builtin.Word, %110 : $Builtin.Int8) // user: %128
  %120 = integer_literal $Builtin.Int64, 840      // user: %121
  %121 = struct $UInt (%120 : $Builtin.Int64)     // user: %128
  %122 = string_literal utf8 "Fatal error"        // user: %124
  %123 = integer_literal $Builtin.Word, 11        // user: %125
  %124 = builtin "ptrtoint_Word"(%122 : $Builtin.RawPointer) : $Builtin.Word // user: %125
  %125 = struct $StaticString (%124 : $Builtin.Word, %123 : $Builtin.Word, %110 : $Builtin.Int8) // user: %128
  %126 = struct $UInt32 (%111 : $Builtin.Int32)   // user: %128
  // function_ref _fatalErrorMessage(_:_:file:line:flags:)
  %127 = function_ref @$Ss18_fatalErrorMessage__4file4line5flagss5NeverOs12StaticStringV_A2HSus6UInt32VtF : $@convention(thin) (StaticString, StaticString, StaticString, UInt, UInt32) -> Never // user: %128
  %128 = apply %127(%125, %119, %115, %121, %126) : $@convention(thin) (StaticString, StaticString, StaticString, UInt, UInt32) -> Never
  unreachable                                     // id: %129
} // end sil function '$S3tmp15testConvolution1x6filter10TensorFlow0E0VySfGAH_AHtF'

---- END OF INPUT FUNCTION ----------
```

Error:
```
/usr/local/google/home/hongm/ssd_part/git/swift-tensorflow-merge/swift/test/TensorFlow/tmp.swift:5:28: error: attribute 'padding' requires a constant argument
  return x.toAccelerator().convolved2D(withFilter: filter.toAccelerator(),
         ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

mhong commented:
@lattner https://bugs.swift.org/browse/SR-8117  has more context

mhong commented:
I found a low-impact short-term fix for this issue: disable (comment out) the small string optimization code path in the 4 lines of code here https://github.com/apple/swift/blob/ad7cbb3a9324ff3ec99063f2174294feae420c28/stdlib/public/core/String.swift#L832. I'll make a patch based on this and do some tests.

@lattner If you have another suggested fix, please let me know.

Commit:
Found a simpler short-term fix to the small string deabstraction
problem (https://bugs.swift.org/browse/SR-8117), by disabling the small string
optimization code path in String.swift. This allowed us to revert changes to various tests.

The remaining limitation is that an empty string literal is not supported. This
support of empty string attrs is not currently needed in any existing ops (as
far as unti test coverage is concerned), and can be added via more short-term
hacks if needed.

The empty string case generates a new SIL code pattern as follows, where the
_StringObject struct, when representing an empty string, no longer points to a
string_literal eventually, and instead sets its second operand (%5 in the
example below) to 0.

  %0 = integer_literal $Builtin.Int64, 0          // user: %5
  %1 = integer_literal $Builtin.Int64, -2305843009213693952 // user: %2
  %2 = struct $UInt (%1 : $Builtin.Int64)         // user: %3
  %3 = value_to_bridge_object %2 : $UInt          // user: %4
  %4 = struct $_StringObject (%3 : $Builtin.BridgeObject) // user: %6
  %5 = struct $UInt (%0 : $Builtin.Int64)         // user: %6
  %6 = struct $_StringGuts (%4 : $_StringObject, %5 : $UInt) // user: %7
  %7 = struct $String (%6 : $_StringGuts)         // users: %49, %49
  %49 = builtin "__tfop_OneShotIterator,container,dataset_factory,output_types,output_shapes,shared_name"(%7 : $String, %8 : $@convention(tensorflow) () -> @owned VariantHandle, %21 : $Array<Float.Type>, %32 : $Array<TensorShape>, %7 : $String) : $ResourceHandle // users: %85, %77, %51, %50
  debug_value %49 : $ResourceHandle, let, name "iterator" // id: %50

mhong commented:
@swift-ci please test tensorflow

mhong commented:
Ok, the only test failing from my local testing is TensorFlow/tensor_autodiff.swift, which @rxwei / @dan-zheng will be looking into shortly.

@rxwei @lattner, PTAL this PR.

rxwei commented:
With the hack in the standard library, are those changes in TFPartition still necessary?

mhong commented:
Which ones? https://github.com/apple/swift/pull/17546/files does not involve changes in TFPartition.{h,cpp} any more.

mhong commented:
https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/249/ confirmed on the test status.

rxwei commented:
Ah, I mistook the change from `device` to `__device` as part of the fix. Understood that it's a separate thing :)

LGTM now.

lattner commented:
I'm ok with disabling the short string optzn for now. That will presumably break some non-TF tests though.  It is a useful way to unblock the merge though!  Thanks!

mhong commented:
@lattner Interestingly, https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/249/ showed the only failed test is auto_diff. I'll go ahead and merge.
